cranelift/assembler-x64/meta/src/generate/format.rs

```diff
@@ -3,6 +3,36 @@
 use super::{Formatter, fmtln};
 use crate::dsl;
 
+/// Different methods of emitting a ModR/M operand and encoding various bits and
+/// pieces of information into it. The REX/VEX formats plus the operand kinds
+/// dictate how exactly each instruction uses this, if at all.
+enum ModRmStyle {
+    /// This instruction does not use a ModR/M byte.
+    None,
+
+    /// The R/M bits are encoded with `reg`, and the Reg/Opcode bits are encoded
+    /// with `digit`.
+    RegWithDigit { reg: dsl::Location, digit: u8 },
+
+    /// The R/M bits are encoded with `mem`, and the Reg/Opcode bits are encoded
+    /// with `digit`.
+    RegMemWithDigit { mem: dsl::Location, digit: u8 },
+
+    /// The R/M bits are encoded with `mem`, and the Reg/Opcode bits are encoded
+    /// with `reg`.
+    RegMem {
+        reg: dsl::Location,
+        mem: dsl::Location,
+    },
+
+    /// The R/M bits are encoded with `reg`, and the Reg/Opcode bits are encoded
+    /// with `dst`.
+    RegReg {
+        reg: dsl::Location,
+        dst: dsl::Location,
+    },
```

Comment:
I wonder if we should merge these two into `RegRm { reg: ..., rm: ... }` since that ends up being the way the manual describes this.